### PR TITLE
fix(scaletest/workspacetraffic): wait for non-zero metrics before cancelling in TestRun

### LINKED ISSUE #9661: test flake: scaletest/workspacetraffic TestRun/PTY
```
    run_test.go:123: read errors: 0
    run_test.go:124: write errors: 0
    run_test.go:125: bytes read total: 1170
    run_test.go:126: bytes written total: 1036
    run_test.go:136: 
        	Error Trace:	/home/runner/actions-runner/_work/coder/coder/scaletest/workspacetraffic/run_test.go:136
        	Error:      	Should not be zero, but was 0
        	Test:       	TestRun/PTY
```

https://github.com/coder/coder/actions/runs/6171775603/job/16750697796?pr=9658

### PR DESCRIPTION
**Pull request by johnstcn**
Hopefully fixes https://github.com/coder/coder/issues/9661

### CODE REVIEW

File: scaletest/workspacetraffic/run_test.go [hunk 1]
```diff
@@ -226,12 +238,32 @@ func TestRun(t *testing.T) {
...
-			<-time.After(cancelAfter)
-			cancel()
+			defer close(runDone)
+			err := runner.Run(ctx, "", &logs)
+			assert.NoError(t, err, "unexpected error calling Run()")
 		}()
-		require.NoError(t, runner.Run(ctx, "", &logs), "unexpected error calling Run()")
+
+		gotMetrics := make(chan struct{})
+		go func() {
```
Comment (mtojek) [inline scaletest/workspacetraffic/run_test.go:250]: nit: I'm always confused if there is a possibility of a routine leak. defer goes first then routine is killed, right?

**Review by mtojek — APPROVED** [review]
Fingers crossed!

File: scaletest/workspacetraffic/run_test.go (hunk 1 above)
Comment (johnstcn) [reply scaletest/workspacetraffic/run_test.go:250]: Defer runs when the surrounding function returns. So `assert.Eventually()` has to finish before `gotMetrics` is closed, and `<-gotMetrics` will block until that happens.

File: scaletest/workspacetraffic/run_test.go [hunk 2]
```diff
@@ -302,3 +328,12 @@ func (m *testMetrics) Latencies() []float64 {
 	defer m.Unlock()
 	return m.latencies
 }
+
+func anyF[T any](vals []T, fn func(T) bool) bool {
```
Comment (mafredri) [inline scaletest/workspacetraffic/run_test.go:332]: I think this is ~`slices.ContainsFunc`? (Just a tip to reduce code.)

**Review by mafredri — APPROVED** [review]

**Pull request merged**